Add support for reading Arrow files

jonmmease opened the pull request:
# Which issue does this PR close?
Closes #5594

# What changes are included in this PR?

This PR adds support for registering Arrow files with the `SessionContext` and reading them. It's largely pattern matched on the existing Avro support, with some inspiration from #1858.

# Are these changes tested?
There is one test, but I'd appreciate guidance on what additional testing we'd like to have here.

# Are there any user-facing changes?
Yes, there is now a `register_arrow` method on the `SessionContext`.

jonmmease commented:
Ok, tests are passing and this is ready for a look. Thanks!

alamb commented:
Thank you @jonmmease  -- this looks great. I plan to review it in the next day or so if no one beats me to it.

alamb commented on datafusion/core/src/datasource/file_format/arrow.rs:
```diff
@@ -0,0 +1,94 @@
+            let schema = match store.get(&object.location).await? {
+                GetResult::File(mut file, _) => read_arrow_schema_from_reader(&mut file)?,
+                r @ GetResult::Stream(_) => {
+                    // TODO: Fetching entire file to get schema is potentially wasteful
```
Yes I agree it is more than potentially 😆 

I don't think we need to fix it for this PR, however. Maybe @tustvold  has some ideas, and if not then I think we can just file a follow on ticket to track fetching only the parts that are needed

jonmmease replied:
Haha, yeah, I carried that comment over from the Avro reader as the situation is the same here. But maybe there's some way to look at the start of the stream and parse out the schema?

alamb commented on datafusion/core/tests/sql/arrow_files.rs:
```diff
@@ -0,0 +1,70 @@
+
+use super::*;
+
+async fn register_arrow(ctx: &mut SessionContext) {
```
Looks good -- I think we should also write a sqllogictest version of this (e.g. https://github.com/apache/arrow-datafusion/blob/b578c5819fc05801f2972dd427fbc13cf4773ea8/datafusion/core/tests/sqllogictests/test_files/ddl.slt#LL259C1-L260C1)

So that we have coverage of creating an arrow backed table via SQL

But I also think that could be added as a follow on PR

jonmmease replied:
registration support added in cfad814d21e51a2a3bbea9dacf6a9dc42977397a, and sqllogictest added in 718a21ffaaef71f71d693f0b6c9d705373c75702

alamb reviewed (approved):
I think this looks great @jonmmease  -- thank you so much 👍 

I think we should add a sqllogictest (I can help with this) if possible to cover SQL as well and file a ticket for the follow on work to avoid reading the file twice, but otherwise I think this PR is good to go.

Really nice work 🏆

jonmmease commented on datafusion/core/src/datasource/file_format/arrow.rs:
```diff
@@ -0,0 +1,94 @@
+// specific language governing permissions and limitations
+// under the License.
+
+//! Apache Arrow format abstractions
```
Done in cfad814d21e51a2a3bbea9dacf6a9dc42977397a

jonmmease commented:
Thanks for taking a look and for the kind works @alamb! I think I've addressed your feedback above. Let me know if you'd like me to open the issue to track the schema-from-stream TODO.

alamb commented:
> Let me know if you'd like me to open the issue to track the schema-from-stream TODO.

Yes, please, if you can that would be awesome 🙏  thank you

alamb reviewed (approved):
Looks great -- thank you @jonmmease 👨‍🍳 👌

jonmmease commented:
Thanks a lot for the helpful review @alamb!  I opened https://github.com/apache/arrow-datafusion/issues/6368 as discussed.
